src/oc/hoc.cpp

```diff
@@ -653,77 +650,34 @@ void hoc_show_errmess_always(void) {
 }
 
 int hoc_execerror_messages;
-
-/* this is possibly non-portable since it is based on the declaration in
-    setjmp.h of
-    typedef int jmp_buf[_JBLEN];
-*/
-void (*oc_jump_target_)(); /* see ivoc/SRC/ocjump.cpp */
-
-
+int nrn_try_catch_nest_depth{0};
 int yystart;
 
-/* what to do about partially constructed objects at hoc_execerror */
-extern void hoc_newobj1_err();
-
-/** If one of the two jmp_buf is controlling the longjmp
- *  hoc_newobj1_err needs handle to know how much to unwrap the newobj1 stack.
- **/
-void* nrn_get_hoc_jmp() {
-    void* jmp = hoc_oc_jmpbuf ? (void*) hoc_oc_begin : (void*) begin;
-    return jmp;
-}
-
 void hoc_execerror_mes(const char* s, const char* t, int prnt) { /* recover from run-time error */
     hoc_in_yyparse = 0;
     yystart = 1;
     hoc_menu_cleanup();
     hoc_errno_check();
-#if 0
-	hoc_xmenu_cleanup();
-#endif
     if (debug_message_ || prnt) {
-        warning(s, t);
+        hoc_warning(s, t);
         frame_debug();
         nrn_err_dialog(s);
-#if defined(__GO32__)
-        {
-            extern int egagrph;
-            if (egagrph) {
-                hoc_outtext("Error:");
-                hoc_outtext(s);
-                if (t) {
-                    hoc_outtext(" ");
-                    hoc_outtext(t);
-                }
-                hoc_outtext("\n");
-            }
-        }
-#endif
-    }
-    /* in case warning not called */
-    ctp = cbuf;
-    *ctp = '\0';
-
-    if (oc_jump_target_ && (nrnmpi_numprocs_world == 1 || !nrn_mpiabort_on_error_)) {
-        hoc_newobj1_err();
-        (*oc_jump_target_)();
     }
+    // In case hoc_warning not called
+    hoc_ctp = hoc_cbuf;
+    *hoc_ctp = '\0';
+    // There used to be some logic here to abort here if we are inside an OcJump call.
 #if NRNMPI
     if (nrnmpi_numprocs_world > 1 && nrn_mpiabort_on_error_) {
         nrnmpi_abort(-1);
     }
 #endif
     hoc_execerror_messages = 1;
-    if (fin && pipeflag == 0 && (!nrn_fw_eq(fin, stdin) || !nrn_istty_))
-        IGNORE(nrn_fw_fseek(fin, 0L, 2)); /* flush rest of file */
-    hoc_oop_initaftererror();
-    if (hoc_oc_jmpbuf) {
-        hoc_newobj1_err();
-        longjmp(hoc_oc_begin, 1);
+    if (hoc_fin && hoc_pipeflag == 0 && (!nrn_fw_eq(hoc_fin, stdin) || !nrn_istty_)) {
+        IGNORE(nrn_fw_fseek(hoc_fin, 0L, 2)); /* flush rest of file */
     }
-    hoc_newobj1_err();
-    longjmp(begin, 1);
+    hoc_oop_initaftererror();
+    throw std::runtime_error("hoc_execerror");
```

Comment:
In future we could actually use the thrown exception to propagate error information.